pkg/controllers/scheduler/framework/plugins/preferencebinpack/planner.go

```diff
@@ -0,0 +1,331 @@
+		return currentPlan, desiredOverflow, nil
+	case currentTotalOkReplicas > desiredTotalReplicas:
+		plan, err := scaleDown(
+			currentPlan, desiredPlan,
+			currentTotalOkReplicas-totalReplicas,
+			availableClusters,
+		)
+		if err != nil {
+			return nil, nil, err
+		}
+		klog.V(4).Infof("ScaleDown plan: %v", plan)
+		return plan, desiredOverflow, nil
+	default:
+		plan, err := scaleUp(
+			rsp,
+			currentPlan, desiredPlan,
+			limitedCapacity,
+			totalReplicas-currentTotalOkReplicas,
+			availableClusters,
+		)
+		if err != nil {
+			return nil, nil, err
+		}
+		klog.V(4).Infof("ScaleUp plan: %v", plan)
+		return plan, desiredOverflow, nil
+	}
+}
+
+func getDesiredPlan(
+	preferences []*namedClusterPreferences,
+	estimatedCapacity map[string]int64,
+	limitedCapacity map[string]int64,
+	totalReplicas int64,
+) (map[string]int64, map[string]int64) {
+	remainingReplicas := totalReplicas
+	plan := make(map[string]int64, len(preferences))
+	overflow := make(map[string]int64, len(preferences))
+
+	// Assign each cluster the minimum number of replicas it requested.
+	for _, preference := range preferences {
+		min := framework.MinInt64(preference.MinReplicas, remainingReplicas)
+		if capacity, hasCapacity := limitedCapacity[preference.clusterName]; hasCapacity && capacity < min {
+			min = capacity
+		}
+		if capacity, hasCapacity := estimatedCapacity[preference.clusterName]; hasCapacity && capacity < min {
+			overflow[preference.clusterName] = min - capacity
+			min = capacity
+		}
+		remainingReplicas -= min
+		plan[preference.clusterName] = min
+	}
+
+	for _, preference := range preferences {
+		start := plan[preference.clusterName]
+
+		// In total there should be the amount that was there at start plus whatever is due
+		// in this iteration
+		total := start + remainingReplicas
+
+		if preference.MaxReplicas != nil && total > *preference.MaxReplicas {
+			total = *preference.MaxReplicas
+		}
+		if capacity, hasCapacity := limitedCapacity[preference.clusterName]; hasCapacity && total > capacity {
+			total = capacity
+		}
+		if capacity, hasCapacity := estimatedCapacity[preference.clusterName]; hasCapacity && total > capacity {
+			overflow[preference.clusterName] += total - capacity
+			total = capacity
+		}
+
+		// Only total-start replicas were actually taken.
+		remainingReplicas -= total - start
+		plan[preference.clusterName] = total
+	}
+
+	return plan, overflow
+}
+
+func scaleUp(
+	rsp *ReplicaSchedulingPreference,
+	currentReplicaCount, desiredReplicaCount map[string]int64,
+	limitedCapacity map[string]int64,
+	scaleUpCount int64,
+	availableClusters []string,
+) (map[string]int64, error) {
+	namedPreferences := make([]*namedClusterPreferences, 0, len(availableClusters))
+	for _, cluster := range availableClusters {
+		current := currentReplicaCount[cluster]
+		desired := desiredReplicaCount[cluster]
+		if desired > current {
+			pref := &namedClusterPreferences{
+				clusterName: cluster,
```

Comment:
Should the weights in the new `pref` equal those in `rsp`? When scaling up, we still want to fill clusters with higher priority.